### ISSUE TITLE
Installing global package will create directory with name '~' in current directory

### ISSUE THREAD

**Issue by sadika9** [yarnpkg/yarn#2008]
When adding global package yarn will create directory with name '~' in current directory.

```
$ mkdir ~/temp && cd ~/temp
$ yarn global add nodemon
```
this will create directory "~" inside ~/temp

Node.js: v7.1.0
yarn: v0.17.8
OS: ArchLinux
kernel: Linux 4.8.8-2

**Comment by DenisGorbachev** [comment]
Can't reproduce on Linux Mint 17.3.

**Comment by lslvxy** [comment]
if you has set `npm config` ,` prefix: '~/.npm-global'`,

then you run `yarn global install ` in `~/temp` will creat dir `~/temp/~/.nmp-global`

**Comment by sadika9** [comment]
@lslvxy  I just noticed that. I have set my prefix to ~/.node_modules in ~/.npmrc file

**Comment by lslvxy** [comment]
you can run `yarn config set prefix ~/.yarn-global` then run `yarn config get prefix ` it show as `/home/user/.yarn-global` not `~/.yarn-global`

is this a bug?

**Comment by bayprogrammer** [comment]
I installed the package on Fedora 25 Workstation using `dnf` (after having enabled the repo per the documentation installation directions). I use the z-shell as my shell.

When I run `yarn global add create-react-app` it created a directory called `~` (the literal tilde, not my home directory) inside of my home directory, which several files in there.

```
$ cd $HOME
$ tree -a \~
~
└── .local
    └── bin
        └── create-react-app -> ../../../.config/yarn/global/node_modules/.bin/create-react-app

2 directories, 1 file
```

`yarn global bin` returns "/home/zebdeos/~/.local/bin". `yarn config get prefix` returns "~/.local".

This happens even after a clean environment on my system where I deleted the yarn trees under `$HOME/.local`, `$HOME/.config`, `$HOME/.cache`, `$HOME/.yarn`, `$HOME/.yarnrc`, verifying no yarn-related files are under `$HOME/.local/{lib,share,bin}`, as well as deleting my old `$HOME/.npm`; I also took care for my testing to ensure system node is in use here (normally I use nodenv).

After the above tests I realized I have an `.npmrc` file which sets my prefix to `~/.local`. I did some searching around. I found a similar sounding issue relating specifically on the npm package bugs list for the Ubuntu package managed on LaunchPad: https://bugs.launchpad.net/ubuntu/+source/npm/+bug/1045837.

Removing `.npmrc` from my home directory and re-running the above tests (again after clearing all the previoulsy mentioned config locations), I get different results. Now `yarn config get prefix` returns `undefined`, `yarn global bin` returns `/usr/bin`, and the output of `yarn config list` under `info npm config` is just `{}` (empty, I presume, because I removed my `.npmrc`).

Attemping `yarn global add create-react-app` gives the following output (including an error about not having permissions to mess with the system `/usr/bin` files):

```
yarn global v0.18.1
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
success Installed create-react-app@1.0.2 with binaries:
      - create-react-app
error We don't have permissions to touch the file "/usr/bin/create-react-app".
info Visit https://yarnpkg.com/en/docs/cli/global for documentation about this command.
```

However, if I first run `yarn config set prefix ~/.local` (per hint from @lslvxy), uninstall the package, then try to reinstall the package (after first uninstalling it to ensure the bin links get recreated), I get the following results:

```
$ yarn global bin
/home/zebdeos/.local/bin
$ yarn config get prefix
/home/zebdeos/.local
$ yarn global remove create-react-app
...
$ yarn global add create-react-app
...
$ which create-react-app
~/.local/bin/create-react-app
$ ls -l ~/.local/bin/create-react-app
lrwxrwxrwx. 1 zebdeos zebdeos 60 Dec 20 22:49 /home/zebdeos/.local/bin/create-react-app -> ../../.config/yarn/global/node_modules/.bin/create-react-app*
```

So it *seems like* Yarn reads the prefix settings in ~/.npmrc, if it exists, but does not attempt to expand the `~` symbol to $HOME; reather it uses the value unexpanded a literal path relative to the current working directory. When running the command I was in my home directory, so the literal `~` directory got created under `$HOME` (I note according to @sadika9's original post the CWD was `$HOME/temp` and the `~` directory created under there).

Setting the config value for prefix directly for yarn via the above commands *does* seem to work as expected, I'm guessing because the shell expansion happens before `yarn` recieves the arguments (so it just sets the prefix to the absolute path to my home directory, which is fine with me).
